solidity/test/rewards/TestRewardUpgrades.js

```diff
@@ -0,0 +1,222 @@
+chai.use(require('bn-chai')(BN))
+const expect = chai.expect
+
+describe.only("Rewards/Upgrades", () => {
+    const owner = accounts[0]
+    const thirdParty = accounts[1]
+    const beneficiary = accounts[3]
+
+    const intervalWeights = [5, 10, 15, 20]
+    const totalRewards = 1000000
+    const minimumKeepsPerInterval = 1
+    const termLength = 1000
+    let timestamps
+    let firstIntervalStart
+
+    let token, rewards, newRewards
+
+    before(async () => {
+        token = await KeepToken.new({ from: owner })
+        newRewards = await NewRewardsStub.new()
+
+        firstIntervalStart = await time.latest()
+        timestamps = [
+            101, 150,    // 2 keep in interval 0
+            1100,        // 1 keep in interval 1
+            2200, 2201   // 2 keeps in interval 2
+        ].map((t) => firstIntervalStart.addn(t).toNumber())
+
+        rewards = await RewardsStub.new(
+            token.address,
+            minimumKeepsPerInterval,
+            firstIntervalStart,
+            intervalWeights,
+            timestamps,
+            termLength,
+            {from: owner}
+         )
+         await token.approveAndCall(
+            rewards.address,
+            totalRewards,
+            "0x0",
+            {from: owner}
+        )
+        await rewards.markAsFunded({from: owner})
+    })
+
+    beforeEach(async () => {
+        await createSnapshot()
+    })
+
+    afterEach(async () => {
+        await restoreSnapshot()
+    })
+
+    describe("upgrades", async () => {
+        it("can be initiated only by contract owner", async () => {            
+            await expectRevert(
+                rewards.initiateRewardsUpgrade(
+                    newRewards.address,
+                    {from: thirdParty}
+                ),
+                "Ownable: caller is not the owner"
+            )
+        })
+
+        it("can be finalized only by contract owner", async () => {
+            await rewards.initiateRewardsUpgrade(
+                newRewards.address,
+                {from: owner}
+            )
+            await expectRevert(
+                rewards.finalizeRewardsUpgrade({from: thirdParty}),
+                "Ownable: caller is not the owner."
+            )            
+        })
+
+        it("can not be finalized without initiating first", async () => {
+            await expectRevert(
+                rewards.finalizeRewardsUpgrade({from: owner}),
+                "Upgrade not initiated"
+            ) 
+        })
+
+        it("can not be finalized before the current interval ends", async () => {
+            await rewards.initiateRewardsUpgrade(
+                newRewards.address,
+                {from: owner}
+            )
+            await expectRevert(
+                rewards.finalizeRewardsUpgrade({from: owner}),
+                "Interval hasn't ended yet"
+            )
+        })
+
+        it("can not be finalized another time without initiating again", async () => {
```

Comment:
another time, you mean in another interval?

Reply:
No, I man calling `finalizeRewardsUpgrade` twice.